File: tesis_gemelo_digital/src/lib/inverterService.ts
import { ObjectId, WithId } from 'mongodb';
import { InverterConfig } from '@/types';
import { getDb } from './db';

const COLLECTION_NAME = 'inversores';

type InverterDocument = Omit<InverterConfig, '_id' | 'createdAt' | 'updatedAt'> & {
  createdAt: Date;
  updatedAt: Date;
};

function mapInverter(doc: WithId<InverterDocument>): InverterConfig {
  return {
    _id: doc._id.toHexString(),
    manufacturer: doc.manufacturer,
    model: doc.model,
    ratedPowerKw: doc.ratedPowerKw,
    quantity: doc.quantity,
    efficiencyPercent: doc.efficiencyPercent,
    createdAt: doc.createdAt.toISOString(),
    updatedAt: doc.updatedAt.toISOString(),
  };
}

function ensurePositive(value: unknown, field: string): number {
  const num = Number(value);
  if (!Number.isFinite(num) || num <= 0) {
    throw new Error(`El campo ${field} debe ser un número mayor que cero.`);
  }
  return num;
}

function ensureNonNegative(value: unknown, field: string): number {
  const num = Number(value);
  if (!Number.isFinite(num) || num < 0) {
    throw new Error(`El campo ${field} debe ser un número mayor o igual que cero.`);
  }
  return num;
}

export async function listInverters(): Promise<InverterConfig[]> {
  const db = await getDb();
  const collection = db.collection<InverterDocument>(COLLECTION_NAME);
  const docs = await collection.find().sort({ updatedAt: -1 }).toArray();
  return docs.map(mapInverter);
}

export async function getInverterById(id: string): Promise<InverterConfig | null> {
  const db = await getDb();
  const collection = db.collection<InverterDocument>(COLLECTION_NAME);
  const _id = new ObjectId(id);
  const doc = await collection.findOne({ _id });
  return doc ? mapInverter(doc) : null;
}

export async function createInverter(payload: Partial<InverterConfig>): Promise<InverterConfig> {
  if (!payload.manufacturer?.trim()) {
    throw new Error('El campo manufacturer es obligatorio.');
  }

  const now = new Date();
  const document: InverterDocument = {
    manufacturer: payload.manufacturer.trim(),
    model: payload.model,
    ratedPowerKw: ensurePositive(payload.ratedPowerKw, 'ratedPowerKw'),
    quantity: Math.trunc(ensurePositive(payload.quantity, 'quantity')),
    efficiencyPercent:
      payload.efficiencyPercent !== undefined
        ? ensureNonNegative(payload.efficiencyPercent, 'efficiencyPercent')
        : undefined,
    createdAt: now,
    updatedAt: now,
  };

  const db = await getDb();
  const collection = db.collection<InverterDocument>(COLLECTION_NAME);
  const result = await collection.insertOne(document);

  return {
    ...payload,
    _id: result.insertedId.toHexString(),
    manufacturer: document.manufacturer,
    model: document.model,
    ratedPowerKw: document.ratedPowerKw,
    quantity: document.quantity,
    efficiencyPercent: document.efficiencyPercent,
    createdAt: document.createdAt.toISOString(),
    updatedAt: document.updatedAt.toISOString(),
  };
}

export async function updateInverter(
  id: string,
  payload: Partial<InverterConfig>
): Promise<InverterConfig | null> {
  const db = await getDb();
  const collection = db.collection<InverterDocument>(COLLECTION_NAME);
  const _id = new ObjectId(id);

  const existing = await collection.findOne({ _id });
  if (!existing) {
    return null;
  }

  const update: Partial<InverterDocument> = {};

  if (payload.manufacturer !== undefined) {
    if (!payload.manufacturer.trim()) {
      throw new Error('El campo manufacturer es obligatorio.');
    }
    update.manufacturer = payload.manufacturer.trim();
  }
  if (payload.model !== undefined) update.model = payload.model;
  if (payload.ratedPowerKw !== undefined) {
    update.ratedPowerKw = ensurePositive(payload.ratedPowerKw, 'ratedPowerKw');
  }
  if (payload.quantity !== undefined) {
    update.quantity = Math.trunc(ensurePositive(payload.quantity, 'quantity'));
  }
  if (payload.efficiencyPercent !== undefined) {
    update.efficiencyPercent = ensureNonNegative(payload.efficiencyPercent, 'efficiencyPercent');
  }

  update.updatedAt = new Date();

  await collection.updateOne({ _id }, { $set: update });
  const refreshed = await collection.findOne({ _id });

  return refreshed ? mapInverter(refreshed) : null;
}

export async function deleteInverter(id: string): Promise<boolean> {
  const db = await getDb();
  const collection = db.collection<InverterDocument>(COLLECTION_NAME);
  const _id = new ObjectId(id);
  const result = await collection.deleteOne({ _id });
  return result.deletedCount === 1;
}

File: tesis_gemelo_digital/backend/app/services/inverter_service.py
"""
Inverter CRUD helpers backed by MongoDB.
"""
from __future__ import annotations

from datetime import datetime
from typing import Any, Dict, List, Optional

from bson import ObjectId
from pymongo.collection import Collection

from app.database import get_database

COLLECTION_NAME = "inversores"


def _collection() -> Collection:
    return get_database()[COLLECTION_NAME]


def _ensure_positive(value: Any, field: str) -> float:
    if value is None:
        raise ValueError(f"El campo {field} es obligatorio.")
    number = float(value)
    if number <= 0:
        raise ValueError(f"El campo {field} debe ser un número mayor que cero.")
    return number


def _ensure_non_negative(value: Any, field: str) -> float:
    if value is None:
        raise ValueError(f"El campo {field} es obligatorio.")
    number = float(value)
    if number < 0:
        raise ValueError(f"El campo {field} debe ser un número mayor o igual que cero.")
    return number


def _ensure_text(value: Any, field: str) -> str:
    if value is None:
        raise ValueError(f"El campo {field} es obligatorio.")
    text = str(value).strip()
    if not text:
        raise ValueError(f"El campo {field} es obligatorio.")
    return text


def _object_id(inverter_id: str) -> ObjectId:
    try:
        return ObjectId(inverter_id)
    except Exception as exc:  # pragma: no cover
        raise ValueError("Identificador de inversor inválido.") from exc


def _map_inverter(doc: Dict[str, Any]) -> Dict[str, Any]:
    return {
        "_id": str(doc["_id"]),
        "manufacturer": doc.get("manufacturer"),
        "model": doc.get("model"),
        "ratedPowerKw": doc.get("ratedPowerKw"),
        "quantity": doc.get("quantity"),
        "efficiencyPercent": doc.get("efficiencyPercent"),
        "createdAt": doc.get("createdAt").isoformat() if doc.get("createdAt") else None,
        "updatedAt": doc.get("updatedAt").isoformat() if doc.get("updatedAt") else None,
    }


def list_inverters() -> List[Dict[str, Any]]:
    cursor = _collection().find().sort("updatedAt", -1)
    return [_map_inverter(doc) for doc in cursor]


def get_inverter(inverter_id: str) -> Optional[Dict[str, Any]]:
    doc = _collection().find_one({"_id": _object_id(inverter_id)})
    return _map_inverter(doc) if doc else None


def create_inverter(payload: Dict[str, Any]) -> Dict[str, Any]:
    manufacturer = _ensure_text(payload.get("manufacturer"), "manufacturer")

    now = datetime.utcnow()
    document: Dict[str, Any] = {
        "manufacturer": manufacturer,
        "model": payload.get("model"),
        "ratedPowerKw": _ensure_positive(payload.get("ratedPowerKw"), "ratedPowerKw"),
        "quantity": int(_ensure_positive(payload.get("quantity"), "quantity")),
        "efficiencyPercent": (
            _ensure_non_negative(payload.get("efficiencyPercent"), "efficiencyPercent")
            if payload.get("efficiencyPercent") is not None
            else None
        ),
        "createdAt": now,
        "updatedAt": now,
    }
    result = _collection().insert_one(document)
    document["_id"] = result.inserted_id
    return _map_inverter(document)


def update_inverter(inverter_id: str, payload: Dict[str, Any]) -> Optional[Dict[str, Any]]:
    update: Dict[str, Any] = {}

    if "manufacturer" in payload:
        update["manufacturer"] = _ensure_text(payload.get("manufacturer"), "manufacturer")
    if "model" in payload:
        update["model"] = payload.get("model")
    if "ratedPowerKw" in payload and payload["ratedPowerKw"] is not None:
        update["ratedPowerKw"] = _ensure_positive(payload["ratedPowerKw"], "ratedPowerKw")
    if "quantity" in payload and payload["quantity"] is not None:
        update["quantity"] = int(_ensure_positive(payload["quantity"], "quantity"))
    if "efficiencyPercent" in payload:
        if payload["efficiencyPercent"] is None:
            update["efficiencyPercent"] = None
        else:
            update["efficiencyPercent"] = _ensure_non_negative(
                payload["efficiencyPercent"],
                "efficiencyPercent",
            )

    if not update:
        return get_inverter(inverter_id)

    update["updatedAt"] = datetime.utcnow()
    result = _collection().find_one_and_update(
        {"_id": _object_id(inverter_id)},
        {"$set": update},
        return_document=True,
    )
    return _map_inverter(result) if result else None


def delete_inverter(inverter_id: str) -> bool:
    result = _collection().delete_one({"_id": _object_id(inverter_id)})
    return result.deleted_count == 1
